File: Classes/Body Parts/Hands.vb
Public Class Hands

    Private CASTKEY As String
    Private LUREKEY As String
    Private CHARMKEY As String
    Private RAFTKEY As String
    Private BAITKEY1 As String
    Private BAITKEY2 As String
    Private BAITKEY3 As String
    Private BAITKEY4 As String
    Private BAITKEY5 As String
    Private BAITKEY6 As String
    Private BAITKEY7 As String
    Private HEARTHKEY As String
    Private BAITKEYS As String()

    Public Sub New()
        UpdateKeys()
    End Sub

    Public Sub UpdateKeys()
        If My.Settings.UseAltKey Then
            CASTKEY = "%(" + My.Settings.FishKey + ")"
            LUREKEY = "%(" + My.Settings.LureKey + ")"
            CHARMKEY = "%(" + My.Settings.CharmKey + ")"
            RAFTKEY = "%(" + My.Settings.RaftKey + ")"
            BAITKEY1 = "%(" + My.Settings.BaitKey1 + ")"
            BAITKEY2 = "%(" + My.Settings.BaitKey2 + ")"
            BAITKEY3 = "%(" + My.Settings.BaitKey3 + ")"
            BAITKEY4 = "%(" + My.Settings.BaitKey4 + ")"
            BAITKEY5 = "%(" + My.Settings.BaitKey5 + ")"
            BAITKEY6 = "%(" + My.Settings.BaitKey6 + ")"
            BAITKEY7 = "%(" + My.Settings.BaitKey7 + ")"
            HEARTHKEY = "%(" + My.Settings.HearthKey + ")"
        Else
            CASTKEY = My.Settings.FishKey
            LUREKEY = My.Settings.LureKey
            CHARMKEY = My.Settings.CharmKey
            RAFTKEY = My.Settings.RaftKey
            HEARTHKEY = My.Settings.HearthKey
            BAITKEY1 = "{" + My.Settings.BaitKey1 + "}"
            BAITKEY2 = "{" + My.Settings.BaitKey2 + "}"
            BAITKEY3 = "{" + My.Settings.BaitKey3 + "}"
            BAITKEY4 = "{" + My.Settings.BaitKey4 + "}"
            BAITKEY5 = "{" + My.Settings.BaitKey5 + "}"
            BAITKEY6 = "{" + My.Settings.BaitKey6 + "}"
            BAITKEY7 = "{" + My.Settings.BaitKey7 + "}"
        End If

        BAITKEYS = {BAITKEY1, BAITKEY2, BAITKEY3, BAITKEY4, BAITKEY5, BAITKEY6, BAITKEY7}
    End Sub

    ''' <summary>
    ''' Performs the 'cast'
    ''' </summary>    
    Public Sub Cast()
        Win32.ActivateWoW()
        Threading.Thread.Sleep(My.Settings.CastingDelay)
        Win32.SendKey(CASTKEY)
    End Sub

    ''' <summary>
    ''' Performs the looting of the fish
    ''' </summary>    
    Public Sub Loot()
        Win32.SendMouseClick()
        Threading.Thread.Sleep(My.Settings.LootingDelay)
    End Sub

    Public Sub ApplyLure()
        Win32.ActivateWoW()
        Win32.SendKey(LUREKEY)
        Threading.Thread.Sleep(3 * 1000)
    End Sub

    Public Sub ApplyCharm()
        Win32.ActivateWoW()
        Win32.SendKey(CHARMKEY)
        Threading.Thread.Sleep(3 * 1000)
    End Sub

    Public Sub ApplyRaft()
        Win32.ActivateWoW()
        Win32.SendKey(RAFTKEY)
        Threading.Thread.Sleep(2 * 1000)
    End Sub

    Public Sub ApplyBait()
        Dim BAITKEY As String = BAITKEY1

        If My.Settings.randomBait Then
            BAITKEY = BAITKEYS(New Random().Next(0, BAITKEYS.Length - 1))
        End If

        Win32.ActivateWoW()
        Win32.SendKey(BAITKEY)
        Threading.Thread.Sleep(2 * 1000)
    End Sub

    Public Sub Hearth()
        Threading.Thread.Sleep(1000)
        Win32.ActivateWoW()
        Win32.SendKey(HEARTHKEY)
    End Sub

End Class

File: UltimateFishBot/Classes/BodyParts/Hands.cs
using System;
using System.Threading;
using System.Threading.Tasks;
using System.Windows.Forms;
using Serilog;
using UltimateFishBot.Helpers;

namespace UltimateFishBot.BodyParts
{
    internal class Hands
    {
        private Cursor _mCursor;
        private int _mBaitIndex;
        private string[] _mBaitKeys;
        private IntPtr _wow;

        private int _aCastingDelay;
        private int _aLootingDelay;

        public Hands(IntPtr wowWindow)
        {
            _wow = wowWindow;
            _mBaitIndex = 0;
            if (Cursor.Current != null) _mCursor = new Cursor(Cursor.Current.Handle);
            UpdateKeys();
        }

        public void SetWow(IntPtr wowWindow) {
            _wow = wowWindow;
        }

        private void UpdateKeys()
        {
            _mBaitKeys = new[]
            {
                Properties.Settings.Default.BaitKey1,
                Properties.Settings.Default.BaitKey2,
                Properties.Settings.Default.BaitKey3,
                Properties.Settings.Default.BaitKey4,
                Properties.Settings.Default.BaitKey5,
                Properties.Settings.Default.BaitKey6,
                Properties.Settings.Default.BaitKey7,
            };
        }

        public async Task Cast(CancellationToken token)
        {
            if (Properties.Settings.Default.RightClickCast)  {
                Win32.SendMouseDblRightClick(_wow);
            } else {
                Win32.SendKey(Properties.Settings.Default.FishKey);
                Log.Information("Sent key: " + Properties.Settings.Default.FishKey);
            }
            var rnd = new Random();
            _aCastingDelay = rnd.Next(Properties.Settings.Default.CastingDelayLow, Properties.Settings.Default.CastingDelayHigh);
            await Task.Delay(_aCastingDelay, token);
        }

        public async Task Loot()
        {
            Win32.SendMouseClick(_wow);
            Log.Information("Send Loot.");
            Random rnd = new Random();
            _aLootingDelay = rnd.Next(Properties.Settings.Default.LootingDelayLow, Properties.Settings.Default.LootingDelayHigh);
            await Task.Delay(_aLootingDelay);
        }

        public void ResetBaitIndex()
        {
            _mBaitIndex = 0;
        }

        public async Task DoAction(NeededAction action, Mouth mouth, CancellationToken cancellationToken)
        {
            string actionKey = "";
            int sleepTime = 0;

            switch (action)
            {
                case NeededAction.HearthStone:
                    {
                        actionKey = Properties.Settings.Default.HearthKey;
                        mouth.Say(Translate.GetTranslate("manager", "LABEL_HEARTHSTONE"));
                        sleepTime = 0;
                        break;
                    }
                case NeededAction.Lure:
                    {
                        actionKey = Properties.Settings.Default.LureKey;
                        mouth.Say(Translate.GetTranslate("manager", "LABEL_APPLY_LURE"));
                        sleepTime = 3;
                        break;
                    }
                case NeededAction.Charm:
                    {
                        actionKey = Properties.Settings.Default.CharmKey;
                        mouth.Say(Translate.GetTranslate("manager", "LABEL_APPLY_CHARM"));
                        sleepTime = 3;
                        break;
                    }
                case NeededAction.Raft:
                    {
                        actionKey = Properties.Settings.Default.RaftKey;
                        mouth.Say(Translate.GetTranslate("manager", "LABEL_APPLY_RAFT"));
                        sleepTime = 2;
                        break;
                    }
                case NeededAction.Bait:
                    {
                        int baitIndex = 0;

                        if (Properties.Settings.Default.CycleThroughBaitList)
                        {
                            if (_mBaitIndex >= 6)
                                _mBaitIndex = 0;

                            baitIndex = _mBaitIndex++;
                        }

                        actionKey = _mBaitKeys[baitIndex];
                        mouth.Say(Translate.GetTranslate("manager", "LABEL_APPLY_BAIT", baitIndex));
                        sleepTime = 3;
                        break;
                    }
                default:
                    return;
            }

            Log.Information("Send key start: " + actionKey);
            Win32.ActivateWow(_wow);
            await Task.Delay(1000, cancellationToken);
            Win32.SendKey(actionKey);
            Log.Information("Sent key: "+actionKey);
            await Task.Delay(sleepTime * 1000, cancellationToken);
        }
    }
}
